apps/nightscout/nightscout.star

```diff
@@ -0,0 +1,250 @@
+                icon = "inputNumeric"
+            ),
+            schema.Text(
+                id = "urgent_high",
+                name = "Urgent High Threshold",
+                desc = "Anything above this is displayed red (Default " + str(DEFAULT_URGENT_HIGH) + ")",
+                icon = "inputNumeric"
+            ),
+            schema.Text(
+                id = "urgent_low",
+                name = "Urgent Low Threshold",
+                desc = "Anything below this is displayed red (Default " + str(DEFAULT_URGENT_LOW) + ")",
+                icon = "inputNumeric"
+            )   
+        ]
+    )
+
+def main(config):
+    nightscout_id = config.get("nightscout_id", None)
+    normal_high = int(config.get("normal_high", DEFAULT_NORMAL_HIGH))
+    normal_low = int(config.get("normal_low", DEFAULT_NORMAL_LOW))
+    urgent_high = int(config.get("urgent_high", DEFAULT_URGENT_HIGH))
+    urgent_low = int(config.get("urgent_low", DEFAULT_URGENT_LOW))
+
+    if nightscout_id == None:
+        print("No Nightscout ID provided - can't do anything")
+        return render.Root(
+            render.Marquee(
+                width = 64,
+                child = render.Text(
+                    content=str("Please specify your Nightscout ID")
+                )
+            )
+        )
+    elif re.match('[^a-zA-Z0-9]', nightscout_id): 
+        # Sanity check the Nightscout ID only consists of Alphanumeric characters
+        print("Invalid Nightscout ID: " + nightscout_id)
+        return render.Root(
+            render.Marquee(
+                width = 64,
+                child = render.Text(
+                    content=str("Nightscout ID contains invalid characters")
+                )
+            )
+        )
+
+
+    # Get the JSON object from the cache
+    nightscout_data_cached = cache.get(nightscout_id + "_nightscout_data")
+    if nightscout_data_cached != None:
+        nightscout_data_json = json.decode(nightscout_data_cached)
+        print("Hit - displaying cached data")
+
+        # Pull the data from the cache
+        sgv_current = int(nightscout_data_json["sgv_current"])
+        sgv_delta = int(nightscout_data_json["sgv_delta"])
+        latest_reading_dt = time.parse_time(nightscout_data_json["latest_reading_date_string"])
+        trend = nightscout_data_json["trend"]
+    else:
+        print("Miss - calling Nightscout API")
+        nightscout_url = "https://" + nightscout_id + ".herokuapp.com/api/v1/entries.json"
+
+        # Request latest entries from the Nightscout URL
+        resp = http.get(nightscout_url)
+
+        if resp.status_code == 502:
+            print("Page not found for nightscout ID " + nightscout_id + " - is this ID correct?")
+            return []
+        elif resp.status_code != 200:
+            fail("Failed to retieve the Nightscout details with status " + str(resp.status_code))
```

Comment:
Given this is infra set up by the user - I think it's best if we never `fail()` in this app and instead always return a message prompting the user that something is wrong with their setup. These `fails()` show up in our error rate that will cause alarms to fire if it's over a certain percentage